clients/spark/core/src/main/scala/io/treeverse/clients/LakeFSInputFormat.scala

```diff
@@ -1,74 +1,145 @@
+    }
     val p = path.toString
     out.writeInt(p.length)
     out.writeChars(p)
   }
 
   override def readFields(in: DataInput): Unit = {
+    byteSize = in.readLong()
+    val rangeIDLength = in.readInt()
+
+    val rangeIDBuilder = new StringBuilder
+    for (_ <- 1 to rangeIDLength) {
+      rangeIDBuilder += in.readChar()
+    }
+    rangeID = rangeIDBuilder.toString
     val encodedRangeLength = in.readInt()
-    val encodedRange = new Array[Byte](encodedRangeLength)
-    in.readFully(encodedRange)
-    range = RangeData.parseFrom(encodedRange)
+    if (encodedRangeLength == 0) {
+      range = null
+    } else {
+      val encodedRange = new Array[Byte](encodedRangeLength)
+      in.readFully(encodedRange)
+      range = RangeData.parseFrom(encodedRange)
+    }
     val pathLength = in.readInt()
     val p = new StringBuilder
     for (_ <- 1 to pathLength) {
       p += in.readChar()
     }
     path = new Path(p.result)
+    logger.debug(s"Read split $this")
   }
-
-  override def getLength: Long = range.estimatedSize
+  override def getLength: Long = byteSize
 
   override def getLocations: Array[String] = Array.empty[String]
 
   override def getLocationInfo: Array[SplitLocationInfo] =
     Array.empty[SplitLocationInfo]
+
+  override def toString: String = {
+    val sb = new StringBuilder
+    sb.append("GravelerSplit: ")
+    sb.append("rangeID: ")
+    sb.append(new String(rangeID))
+    sb.append("\n")
+    sb.append("path: ")
+    sb.append(path)
+    sb.append("\n")
+    sb.append("range: ")
+    sb.append(range)
+    sb.toString
+  }
 }
 
 class WithIdentifier[Proto <: GeneratedMessage with scalapb.Message[Proto]](
     val id: Array[Byte],
-    val message: Proto
+    val message: Proto,
+    val rangeID: String
 ) {}
 
+object EntryRecordReader {
+  val logger: Logger = LoggerFactory.getLogger(getClass.toString)
+}
 class EntryRecordReader[Proto <: GeneratedMessage with scalapb.Message[Proto]](
     companion: GeneratedMessageCompanion[Proto]
 ) extends RecordReader[Array[Byte], WithIdentifier[Proto]] {
+  import EntryRecordReader._
   var it: SSTableIterator[Proto] = _
   var item: Item[Proto] = _
-
+  var gravelerSplit: GravelerSplit = _
   override def initialize(split: InputSplit, context: TaskAttemptContext): Unit = {
     val localFile = File.createTempFile("lakefs.", ".range")
     localFile.deleteOnExit()
-    val gravelerSplit = split.asInstanceOf[GravelerSplit]
+    gravelerSplit = split.asInstanceOf[GravelerSplit]
+
     val fs = gravelerSplit.path.getFileSystem(context.getConfiguration)
     fs.copyToLocalFile(gravelerSplit.path, new Path(localFile.getAbsolutePath))
     // TODO(johnnyaug) should we cache this?
     val sstableReader =
       new SSTableReader(localFile.getAbsolutePath, companion)
+    val props = sstableReader.getProperties()
+    logger.debug(s"Props: $props")
+    if (gravelerSplit.range == null) {
+      if (new String(props.get("type").get) != "ranges" || props.get("entity").nonEmpty) {
```

Comment:
Not sure I understand what this is doing.  In particular, when does this happen?  If it happens, didn't something go badly wrong -- something we'd want to report?

Reply:
This prefix will also include metaranges, and may potentially include "ref-dumps" which represent other entities. This check filters out only ranges.